File: survival-shooter-unity-demo/Assets/rudder-sdk/RudderClient.cs
using com.rudderlabs.unity.library.Event;

namespace com.rudderlabs.unity.library
{
    public class RudderClient
    {
        // singleton instnace of the client
        private static RudderClient instance;
        // instance for event repository (to be used internally)
        private static EventRepository repository;

        private RudderClient() {
            // prevent instance creation through constructor
        }

        // instance initialization method
        public static RudderClient GetInstance(string writeKey)
        {
            return GetInstance(writeKey, Constants.BASE_URL, Constants.FLUSH_QUEUE_SIZE, false);
        }

        // instance initialization method
        public static RudderClient GetInstance(string writeKey, int flushQueueSize)
        {
            return GetInstance(writeKey, Constants.BASE_URL, flushQueueSize, false);
        }

        // instance initialization method
        public static RudderClient GetInstance(string writeKey, string endPointUri)
        {
            return GetInstance(writeKey, endPointUri, Constants.FLUSH_QUEUE_SIZE, false);
        }

        // instance initialization method
        public static RudderClient GetInstance(string writeKey, string endPointUri, int flushQueueSize) {
            return GetInstance(writeKey, endPointUri, flushQueueSize, false);
        }

        // instance initialization method
        public static RudderClient GetInstance(string writeKey, string endPointUri, int flushQueueSize, bool loggingEnabled)
        {
            if (instance == null)
            {
                instance = new RudderClient();

                repository = new EventRepository(writeKey, flushQueueSize, endPointUri, loggingEnabled);
            }
            return instance;
        }

        // getter & setter for endPointUri
        public string GetEndPointUri()
        {
            return EventRepository.endPointUri;
        }
        public void SetEndPointUri(string _endPointUri)
        {
            EventRepository.endPointUri = _endPointUri;
        }

        // getter & setter for flushQueueSize
        public int GetFlushQueueSize()
        {
            return EventRepository.flushQueueSize;
        }
        public void SetFlushQueueSize(int _flushQueueSize)
        {
            EventRepository.flushQueueSize = _flushQueueSize;
        }

        // end point for track events
        public void Track(RudderEvent rudderEvent)
        {
            rudderEvent.message.type = RudderEventType.TRACK.value;
            repository.Dump(rudderEvent);
        }
        public void Track(RudderEventBuilder builder)
        {
            this.Track(builder.Build());
        }

        // end point for page events
        public void Page(RudderEvent rudderEvent)
        {
            rudderEvent.message.type = RudderEventType.PAGE.value;
            repository.Dump(rudderEvent);
        }
        public void Page(RudderEventBuilder builder)
        {
            this.Page(builder.Build());
        }

        // end point for screen events
        public void Screen(RudderEvent rudderEvent)
        {
            rudderEvent.message.type = RudderEventType.PAGE.value;
            repository.Dump(rudderEvent);
        }
        public void Screen(RudderEventBuilder builder)
        {
            this.Screen(builder.Build());
        }

        // end point for identify calls
        public void Identify(RudderTraits rudderTraits)
        {
            RudderEvent rudderEvent = new RudderEventBuilder()
                .SetEventName("Identify")
                .SetUseId(rudderTraits.id)
                .Build();
            rudderEvent.message.type = RudderEventType.IDENTIFY.value;
            rudderEvent.message.context.traits = rudderTraits;
            repository.Dump(rudderEvent);
        }
        public void Identify(RudderTraitsBuilder builder)
        {
            Identify(builder.Build());
        }

        // end point for flushing events
        public void Flush()
        {
            repository.FlushEventsAsync();
        }
    }
}

File: old-references/rudder-client-android-kotlin-old/src/main/java/com/rudderlabs/android/library/RudderClient.kt
package com.rudderlabs.android.library

import android.app.Application
import android.content.Context
import android.util.Log
import com.google.gson.Gson
import com.rudderlabs.android.library.models.RudderEvent
import com.rudderlabs.android.library.models.RudderEventBuilder
import com.rudderlabs.android.library.models.RudderTraits
import com.rudderlabs.android.library.models.porperties.RudderTraitsBuilder
import com.rudderlabs.android.library.models.porperties.validateFor
import com.rudderlabs.android.library.repository.EventRepository
import com.rudderlabs.android.library.util.Constants

object RudderClient {
    private var instance: RudderClient? = null
    private var applicationContext: Application? = null
    private val repository: EventRepository by lazy {
        EventRepository()
    }

    /*
    * get instance api
    * */
    @JvmStatic
    internal fun getClient(): RudderClient {
        if (instance == null) throw IllegalStateException("Client is not initialized")
        else return instance!!
    }

    @JvmStatic
    internal fun getApplicationContext(): Application? {
        return applicationContext
    }

    @JvmStatic
    @JvmOverloads
    fun getInstance(
            context: Context,
            flushQueueSize: Int = 10,
            endPointUri: String = Constants.BASE_URL,
            shouldCache: Boolean = true
    ): RudderClient {
        if (instance == null) {
            if (context.applicationContext == null)
                throw NullPointerException("Application context can not be null")

            applicationContext = context.applicationContext as Application

            repository.initiate(applicationContext!!, shouldCache)

            setFlushQueueSize(flushQueueSize)
            setEndPointUri(endPointUri)

            instance = this
        }
        return instance!!
    }


    /*
    * api to get end point uri
    * */
    fun getEndPointUri() = repository.getEndPointUri()

    /*
    * api to get current flush queue size
    * */
    fun getFlushQueueSize() = repository.getFlushQueueSize()

    /*
    * api to set end point uri
    * */
    fun setEndPointUri(endPointUri: String) {
        repository.setEndPointUri(endPointUri)
    }

    /*
    * api to set flush queue size
    * */
    fun setFlushQueueSize(flushQueueSize: Int) {
        repository.setFlushQueueSize(flushQueueSize)
    }

    /*
    * api to send track event
    * */
    fun track(event: RudderEvent?) {
        Log.d("PostEvent", "")
        if (BuildConfig.DEBUG) {
            println("EVENT: ${Gson().toJson(event)}")
        }
        if (event?.message != null) {
            event.message?.let {
                it.validateFor(EventType.TRACK)
                repository.dump(event.also { e -> e.message?.type = EventType.TRACK.value })
            }
        }
    }

    /*
    * api to send page event
    * */
    fun page(event: RudderEvent?) {
        Log.d("PostEvent", "")
        if (event?.message != null) {
            event.message?.let {
                it.validateFor(EventType.PAGE)
                repository.dump(event.also { e -> e.message?.type = EventType.PAGE.value })
            }
        }
    }

    /*
    * api to send screen event
    * */
    fun screen(event: RudderEvent?) {
        Log.d("PostEvent", "")
        if (event?.message != null) {
            event.message?.let {
                it.validateFor(EventType.SCREEN)
                repository.dump(event.also { event -> event.message?.type = EventType.SCREEN.value })
            }
        }
    }

    /*
    * api to fire identify event (traits)
    * */
    fun identify(traits: RudderTraits) {
        repository.dump(
                RudderEventBuilder()
                        .setChannel("Identification Channel")
                        .setEvent("Identify")
                        .setUserId(traits.id!!)
                        .build().also {
                            it.message?.type = EventType.IDENTIFY.value
                            it.message?.context?.traits = traits
                        }
        )
    }

    fun identify(traitsBuilder: RudderTraitsBuilder) {
        this.identify(traitsBuilder.build())
    }

    /*
    * api to flush events regardless of event count
    * */
    fun flush() {
        repository.flush()
    }
}

